### PR TITLE
feat(coverage): add option to ignore directories and files from coverage report

### PR DESCRIPTION
**Pull request by zerosnacks**
## Motivation

Follow-up of https://github.com/foundry-rs/foundry/pull/7301 as I'm not able to push directly to the fork

This PR includes the commits of https://github.com/dimazhornyk

Closes: #2988 

## Solution

Contrary to https://github.com/foundry-rs/foundry/pull/7301 I decided that for flexibility it is preferable to generalize beyond `glob` and instead use a `regex` on the filepath (not just file name)

- Fixes the formatting issues that slipped in 
- Fixes merge conflicts
- Adds test verifying the applied filter (`contractA` is filtered out, `contractB` is preserved)

This allows for a syntax like:

`forge coverage --no-match-coverage "(script|test|Foo|Bar)"` where `script|test` are obviously paths and `Foo|Bar` files where a glob would be limited to a single folder and not individual files

There is currently no warning applied when all sources are filtered out, that could be a follow up nice to have.

### CODE REVIEW

File: crates/forge/bin/cmd/test/filter.rs [hunk 1]
```diff
@@ -38,6 +38,10 @@ pub struct FilterArgs {
         value_name = "GLOB"
     )]
     pub path_pattern_inverse: Option<GlobMatcher>,
+
+    /// Only show coverage for files that do not match the specified regex pattern.
+    #[arg(long = "no-match-coverage", visible_alias = "nmco", value_name = "REGEX")]
+    pub coverage_pattern_inverse: Option<regex::Regex>,
```
Comment (zerosnacks) [inline crates/forge/bin/cmd/test/filter.rs:44]: I've opted for `no-match-coverage` to stay in line with the `no-match-*` pattern established (`no-match-test`, `no-match-contract`, `no-match-path`) 

See: https://github.com/foundry-rs/foundry/pull/7301#discussion_r1521015253

File: crates/evm/coverage/src/lib.rs [hunk 2]
```diff
@@ -150,6 +150,30 @@ impl CoverageReport {
         }
         Ok(())
     }
+
+    /// Removes all the coverage items that should be ignored by the filter.
+    ///
+    /// This function should only be called after all the sources were used, otherwise, the output
+    /// will be missing the ones that are dependent on them.
+    pub fn filter_out_ignored_sources(&mut self, filter: &impl CoverageFilter) {
```
Comment (mattsse) [inline crates/evm/coverage/src/lib.rs:158]: let's just replace this with an `Fn(&Path) -> bool`

File: crates/evm/coverage/src/lib.rs [hunk 3]
```diff
@@ -150,6 +150,30 @@ impl CoverageReport {
...
+                        .map(|path| filter.matches_file_path(path))
+                        .unwrap_or(false)
+                })
+                .cloned()
+                .collect();
+            if !new_items_for_version.is_empty() {
+                new_items.insert(version.clone(), new_items_for_version);
+            }
+        }
+        self.items = new_items;
```
Comment (mattsse) [inline crates/evm/coverage/src/lib.rs:175]: isn't this just a `HashMap::retain`?

**Review by mattsse — CHANGES_REQUESTED** [review]

**Review by mattsse — APPROVED** [review]
lgtm 

pending @klkvr

**Review by klkvr — APPROVED** [review]
lgtm

**Comment by gosuto-inzasheru** [comment]
thanks for this 🥇

**Pull request merged**